src/account.ts

```diff
@@ -0,0 +1,843 @@
+        } catch (err) {
+          this.master._log.trace(`Failed to forward PREPARE: ${err.message}`)
+          throw new IlpPacket.Errors.InsufficientLiquidityError(err.message)
+        }
+
+        let timer: NodeJS.Timer
+        let response: Buffer = await Promise.race([
+          // Send reject if PREPARE expires before a response is received
+          new Promise<Buffer>(resolve => {
+            timer = setTimeout(() => {
+              resolve(
+                // Opinion: the triggeredBy address isn't useful because upstream connectors can modify
+                // it without any way for downstream nodes to know (even if we trust our direct peers)
+                IlpPacket.errorToReject('', {
+                  ilpErrorCode: 'R00',
+                  message: `Expired at ${new Date().toISOString()}`
+                })
+              )
+            }, expiresAt.getTime() - Date.now())
+          }),
+          // Forward the packet to data handler, wait for response
+          dataHandler(ilp.data)
+        ])
+        // tslint:disable-next-line:no-unnecessary-type-assertion
+        clearTimeout(timer!)
+
+        if (response[0] === IlpPacket.Type.TYPE_ILP_REJECT) {
+          this.subBalance(amountBN)
+        } else if (response[0] === IlpPacket.Type.TYPE_ILP_FULFILL) {
+          this.master._log.trace(`Received FULFILL from data handler in response to forwarded PREPARE`)
+        }
+
+        return ilpAndCustomToProtocolData({ ilp: response })
+      } catch (err) {
+        return ilpAndCustomToProtocolData({ ilp: IlpPacket.errorToReject('', err) })
+      }
+    }
+
+    return []
+  }
+
+  async handleMoney (message: BtpPacket, moneyHandler?: MoneyHandler): Promise<BtpSubProtocol[]> {
+    return this.incomingSettlements.runExclusive(async () => {
+      try {
+        const machinomy = getSubprotocol(message, 'machinomy')
+
+        if (machinomy) {
+          this.master._log.trace(`Handling Machinomy claim for account ${this.account.accountName}`)
+
+          const claim = JSON.parse(machinomy.data.toString())
+
+          const hasValidSchema = claim
+            && typeof claim.value === 'string'
+            && typeof claim.channelId === 'string'
+            && typeof claim.signature === 'string'
+          if (!hasValidSchema) {
+            throw new Error(`claim schema is malformed`)
+          }
+
+          const oldClaim = this.account.bestIncomingClaim
+
+          // If the account already has a channel, client must be using that channel
+          // Calculating profitability across multiple incoming channels sounds like a receipe for disaster
+          if (oldClaim && oldClaim.channelId !== claim.channelId) {
+            throw new Error(`account ${this.account.accountName} must use channel ${oldClaim.channelId}`)
+          }
+
+          // Always fetch updated channel state from network
+          const channel = await fetchChannel(this.master._web3, claim.channelId)
```

Comment:
Does it really make sense to make that API call every time you get a claim in? If the channels are unidirectional you shouldn't need to do that unless you think they might have put more money in the channel and you want to check if they did

Reply:
Maybe not, but it might be a premature optimization. I think it's safer to do it each claim. For example, if the channel was claimed outside of the plugin and the state was not fetched, new claims would still be credited.